pkg/testutils/lint/passes/deferunlockcheck/deferunlockcheck.go

```diff
@@ -0,0 +1,313 @@
+	lt.locks = append(lt.locks, &LockInfo{
+		isRead:           isRead,
+		foundUnlockMatch: false,
+	})
+	ast.Inspect(sel, func(n ast.Node) bool {
+		ident, ok := n.(*ast.Ident)
+		if !ok {
+			return true
+		}
+		// If we find Lock or RLock set the line and we are done
+		// traversing.
+		if ident.Name == "Lock" || ident.Name == "RLock" {
+			position := lt.pass.Fset.Position(n.Pos())
+			lt.locks[len(lt.locks)-1].lineNum = position.Line
+			return false
+		}
+		// Otherwise its part of the prefix that we'll add.
+		lt.locks[len(lt.locks)-1].prefix += ident.Name + "."
+		// Keep traversing if we didn't find an *ast.Ident node.
+		return true
+	})
+}
+
+// maybeReportUnlock tries to find a matching lock for a given unlock by
+// iterating backwards in the locks slice. If one is found, it checks if the
+// distance between is greater than maxLineDistance and also has no nlint
+// comment and reports on that if both are true.
+func (lt *LockTracker) maybeReportUnlock(call *ast.CallExpr, isRead bool) {
+	sel, ok := call.Fun.(*ast.SelectorExpr)
+	if !ok {
+		return
+	}
+	var node ast.Node
+	var unlockPrefix string
+	var unlockLineNum int
+	var unlockPos token.Pos
+	ast.Inspect(sel, func(n ast.Node) bool {
+		ident, ok := n.(*ast.Ident)
+		if !ok {
+			return true
+		}
+		// If we find Unlock or RUnlock set the line and we are done
+		// traversing.
+		if ident.Name == "Unlock" || ident.Name == "RUnlock" {
+			position := lt.pass.Fset.Position(n.Pos())
+			unlockPos = n.Pos()
+			unlockLineNum = position.Line
+			node = n
+			return false
+		}
+		// Otherwise its part of the prefix that we'll add.
+		unlockPrefix += ident.Name + "."
+		// Keep traversing if we didn't find an *ast.Ident node.
+		return true
+	})
+	// Reverse iterate through the locks slice searching for a match.
+	for i := len(lt.locks) - 1; i >= 0; i-- {
+		lock := lt.locks[i]
+		if lock.foundUnlockMatch {
+			continue
+		}
+		// It's only an issue if prefixes match, the lock type matches, distance between is >5 lines away and
+		// there is no nolint comment.
+		if unlockPrefix == lock.prefix && isRead == lock.isRead {
+			lockDistance := unlockLineNum - lock.lineNum
+			if lockDistance > maxLineDistance && !passesutil.HasNolintComment(lt.pass, node, noLintName) {
+				lt.issues = append(lt.issues, analysis.Diagnostic{
+					Pos:     unlockPos,
+					Message: fmt.Sprintf("{RW}Mutex {R}Unlock is >%d lines away from matching {R}Lock", maxLineDistance),
+				})
+			}
+			// This should set on close enough unlocks, nolint unlocks and defer unlocks.
+			lock.foundUnlockMatch = true
+		}
+	}
+}
+
+// shouldReportNonLinearControlFlow checks if locks has a *LockInfo without a matching unlock
+// and if the node has a nolint comment near it. If it finds both it returns true.
+func (lt *LockTracker) shouldReportNonLinearControlFlow(node ast.Node) bool {
+	hasNoLintComment := passesutil.HasNolintComment(lt.pass, node, noLintName)
```

Comment:
Seems like you can immediately `if hasNoLintComment { return false }`, right? We will always return false in this case with the way this function is currently written.